gc.c

```diff
@@ -3587,12 +3853,32 @@ rb_gc_vm_weak_table_foreach(vm_table_foreach_callback_func callback,
         }
         break;
       }
-      case RB_GC_VM_GENERIC_IV_TABLE: {
-        st_table *generic_iv_tbl = rb_generic_ivtbl_get();
-        if (generic_iv_tbl) {
+      case RB_GC_VM_ID_TO_OBJ_TABLE: {
+        if (id_to_obj_tbl) {
             st_foreach(
-                generic_iv_tbl,
-                vm_weak_table_gen_ivar_foreach,
+                id_to_obj_tbl,
+                vm_weak_table_id_to_obj_foreach,
+                (st_data_t)&foreach_data
+            );
+        }
+        break;
+      }
+      case RB_GC_VM_ID_TO_OBJ_TABLE_KEYS: {
+        if (id_to_obj_tbl && !RB_POSFIXABLE(next_object_id)) {
+            st_foreach(
+                id_to_obj_tbl,
+                vm_weak_table_id_to_obj_keys_foreach,
+                (st_data_t)&foreach_data
+            );
+        }
+        break;
+      }
```

Comment:
I think we can merge these two cases together and only pass the key if `RB_POSFIXABLE(next_object_id)`.

Reply:
Good point.